### PR TITLE
Makke wait work in a more predictable case, also allowing tags to be …

### LINKED ISSUE #4078: ec2_ami not handling wait:no and tags correctly
##### ISSUE TYPE
- Bug Report
##### COMPONENT NAME

ec2_ami.py
##### ANSIBLE VERSION

```
ansible 2.1.0.0
```
##### CONFIGURATION

N/A
##### OS / ENVIRONMENT

N/A
##### SUMMARY

I have three problems with the current execution:
1. wait: no still waits. It still waits till the AMI is available or a time_out.
2. As shown below, the tags are present but not displayed as output.
3. When the wait time expires, the AMI is created however the tags are ignored, suggesting they are only added after the wait has expired. I would expect this to happen before on wait: no
##### STEPS TO REPRODUCE

Playbook:

```
    - name: create AMI backup
      ec2_ami:
        region: "{{ ec2_region }}"
        instance_id: "{{ ec2_id }}"
        wait: no
        no_reboot: yes
        name: "{{ ec2_tag_Name }}"
        tags:
          creation: "{{ ansible_date_time.epoch }}"
          expiration: "{{ expiration_date.stdout }}"
      register: output
```
##### EXPECTED RESULTS

I expected the following results:
1. The playbook moves on and does not wait for results
2. Tags are shown in the output
3. Tags show in the AWS console after a time_out has been reached.
##### ACTUAL RESULTS

Output on a normal run (after waiting for the task to complete):

```
TASK [create AMI backup] *******************************************************
changed: [*.*.*.*]

TASK [debug] *******************************************************************
ok: [*.*.*.*] => {
    "output": {
        "architecture": "x86_64",
        "block_device_mapping": {
            "/dev/sda1": {
                "delete_on_termination": true,
                "encrypted": false,
                "size": 8,
                "snapshot_id": "snap-*******",
                "volume_type": "gp2"
            }
        },
        "changed": true,
        "creationDate": "2016-06-29T09:31:56.000Z",
        "description": null,
        "hypervisor": "xen",
        "image_id": "ami-******",
        "is_public": false,
        "location": "*******/*****-2016-06-monthly",
        "msg": "AMI creation operation complete",
        "ownerId": "*******",
        "root_device_name": "/dev/sda1",
        "root_device_type": "ebs",
        "state": "available",
        "tags": {},
        "virtualization_type": "hvm"
    }
}
```

In the situation where a time_out occurs:

```
09:12:36            TASK [create AMI backup] *******************************************************
09:28:18            fatal: [*.*.*.* -> localhost]: FAILED! => {"changed": false, "failed": true, "msg": "Error while trying to find the new image. Using wait=yes and/or a longer wait_timeout may help."}
```

Also notice how this is by no means the 300 seconds ( 5 minutes ) specified as the default.

I think this module needs a good once over to verify the combination of delegation, tags and wait functions properly. Thanks!

### PR DESCRIPTION
**Pull request by jjshoe**
##### ISSUE TYPE

- Bugfix Pull Request
##### COMPONENT NAME

ec2_ami
##### SUMMARY

Fixes #4078

### CODE REVIEW

**Comment by jjshoe** [comment]
@mliudev @hyperized - can I get you guys to test this?

**Comment by gregdek** [comment]
Thanks @jjshoe for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by mliudev** [comment]
Thanks, I'll test this within the next couple of days.

**Comment by jjshoe** [comment]
needsinfo

**Comment by hyperized** [comment]
After applying the patch and trying to create an AMI I get:

`fatal: [10.36.32.123 -> localhost]: FAILED! => {"changed": false, "failed": true, "msg": "Error while trying to find the new image. Using wait=yes and/or a longer wait_timeout may help."}`

The AMI was created.

**Comment by jjshoe** [comment]
@hyperized - I'm not able to reproduce your issue at all. Can you consistently reproduce? If so, can you provide me with the task call you're using?

@mliudev - Had a chance to test?

**Comment by gregdek** [comment]
@jjshoe A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by mliudev** [comment]
@jjshoe Sorry for the slow response. I've recently left the job where I was working on a project that used this module, so I'm no longer in a position to test this. I'll see if I can get the person who's now responsible for that project to take a look.

**Comment by gregdek** [comment]
@jjshoe Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@jjshoe A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by ericamador** [comment]
@jjshoe Thanks for making this PR. @mliudev asked me to take over the testing of this for him. I'm hitting the same thing that @hyperized referenced in his comment above, and the cause seems pretty clear. I'll point the offending line out in a review comment.

File: cloud/amazon/ec2_ami.py [hunk 1]
```diff
@@ -393,7 +393,8 @@ def create_image(module, ec2):
         try:
             img = ec2.get_image(image_id)
 
-            if img.state == 'available':
+            # We must wait for the ami to exist in AWS api responses before we can skip the wait, and add tags when specifying wait: no 
+            if img.state == 'available' or (img and not wait):
```
Comment (ericamador) [inline cloud/amazon/ec2_ami.py:397]: The reason why we appear to getting the "Error while trying to find the new image..." error is that while the change to this line allows for breaking out of the for loop when `img.state` is not `available`, there is a check right immediately after this `for` loop on line 406 where the module fails with this error message `if img.state != 'available'`.

Comment (ericamador) [reply cloud/amazon/ec2_ami.py:397]: I should mention that if I remove that check that begins on line 406, this works well for me. Thanks!

**Pull request closed without merging**